### PR TITLE
feat: add an API to get an allocation's exit status

### PR DESCRIPTION
**Pull request by salonig23**
## Description
Added a get API endpoint to fetch the exit status for each allocation

## Test Plan
Check if endpoints show up correctly in swagger UI

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

**Comment by salonig23** [comment]
> @salonig23 should we review this? Or should we wait till the implementation is added?

Hi Nick! Please review it! I will add the implementation in a separate PR soon!

File: master/internal/api_tasks.go [hunk 1]
```diff
@@ -40,6 +40,9 @@ var (
 	taskLogsFieldsBatchWaitTime = 5 * time.Second
 )
 
+// UnimplementedError is the error return by API endpoints that are not yet implemented.
+var UnimplementedError = status.Error(codes.Unimplemented, "method not yet available")
```
Comment (NicholasBlaskey) [inline master/internal/api_tasks.go:44]: I think we might be able to use ```grpcutil.UnimplementedError```?

File: proto/src/determined/api/v1/api.proto [hunk 2]
```diff
@@ -743,6 +743,15 @@ service Determined {
       tags: "Internal"
     };
   }
+  // Get details about an Allocation.
+  rpc GetAllocation(GetAllocationRequest) returns (GetAllocationResponse) {
+    option (google.api.http) = {
+      get: "/api/v1/allocations/{allocation_id}"
+    };
+    option (grpc.gateway.protoc_gen_swagger.options.openapiv2_operation) = {
+      tags: "Cluster"
```
Comment (NicholasBlaskey) [inline proto/src/determined/api/v1/api.proto:752]: let's tag this as internal for now

File: proto/src/determined/api/v1/trial.proto [hunk 3]
```diff
@@ -520,6 +520,30 @@ message AllocationRendezvousInfoResponse {
...
+  option (grpc.gateway.protoc_gen_swagger.options.openapiv2_schema) = {
+    json_schema: { required: [ "allocation_id" ] }
+  };
+  // The id of the allocation.
+  string allocation_id = 1;
+}
+// Response to GetAllocationRequest
+message GetAllocationResponse {
+  option (grpc.gateway.protoc_gen_swagger.options.openapiv2_schema) = {
+    json_schema: { required: [ "allocation_id", "accelerator_data" ] }
```
Comment (NicholasBlaskey) [inline proto/src/determined/api/v1/trial.proto:534]: I think the fields look wrong in required?

File: proto/src/determined/api/v1/trial.proto [hunk 4]
```diff
@@ -520,6 +520,30 @@ message AllocationRendezvousInfoResponse {
...
+  // The id of the allocation.
+  string allocation_id = 1;
+  // The state of the allocation.
+  string state = 2;
+  // The number of slotss associated with the allocation.
+  int32 slots = 3;
+  // The exit reason for the allocation.
+  string exit_reason = 4;
+  // The status code the allocation exits with.
+  string status_code = 5;
```
Comment (NicholasBlaskey) [inline proto/src/determined/api/v1/trial.proto:545]: can ``status_code`` ever be not an integer in our system?

File: proto/src/determined/api/v1/trial.proto [hunk 5]
```diff
@@ -520,6 +520,30 @@ message AllocationRendezvousInfoResponse {
...
+// Response to GetAllocationRequest
+message GetAllocationResponse {
+  option (grpc.gateway.protoc_gen_swagger.options.openapiv2_schema) = {
+    json_schema: { required: [ "allocation_id", "accelerator_data" ] }
+  };
+  // The id of the allocation.
+  string allocation_id = 1;
+  // The state of the allocation.
+  string state = 2;
+  // The number of slotss associated with the allocation.
```
Comment (NicholasBlaskey) [inline proto/src/determined/api/v1/trial.proto:540]:
```suggestion
  // The number of slots associated with the allocation.
```

File: proto/src/determined/api/v1/trial.proto [hunk 6]
```diff
@@ -520,6 +520,30 @@ message AllocationRendezvousInfoResponse {
...
+  int32 slots = 3;
+  // The exit reason for the allocation.
+  string exit_reason = 4;
```
Comment (NicholasBlaskey) [inline proto/src/determined/api/v1/trial.proto:543]: Can we make exit_reason and status_code optional with 

```optional exit_reason = 4```

**Review by NicholasBlaskey — COMMENTED** [review]
Looks good, couple minor comments

File: master/internal/api_tasks.go (hunk 1 above)
Comment (salonig23) [reply master/internal/api_tasks.go]: Done!

File: proto/src/determined/api/v1/trial.proto (hunk 3 above)
Comment (salonig23) [reply proto/src/determined/api/v1/trial.proto]: Done!

File: proto/src/determined/api/v1/trial.proto (hunk 5 above)
Comment (salonig23) [reply proto/src/determined/api/v1/trial.proto]: Done

File: proto/src/determined/api/v1/trial.proto (hunk 6 above)
Comment (salonig23) [reply proto/src/determined/api/v1/trial.proto]: Done!

File: proto/src/determined/api/v1/trial.proto (hunk 4 above)
Comment (salonig23) [reply proto/src/determined/api/v1/trial.proto]: I don't think so. Changed it to int32

File: proto/src/determined/api/v1/trial.proto [hunk 7]
```diff
@@ -520,6 +520,30 @@ message AllocationRendezvousInfoResponse {
...
+}
+// Response to GetAllocationRequest
+message GetAllocationResponse {
+  option (grpc.gateway.protoc_gen_swagger.options.openapiv2_schema) = {
+    json_schema: { required: [ "allocation_id", "state", "slots" ] }
+  };
+  // The id of the allocation.
+  string allocation_id = 1;
+  // The state of the allocation.
+  string state = 2;
```
Comment (NicholasBlaskey) [inline proto/src/determined/api/v1/trial.proto:539]: can we make state an enum?

https://github.com/determined-ai/determined/blob/928ba32f0e012ca2549a5b1c68d6d10e6e2e5e11/proto/src/determined/task/v1/task.proto#L11

I don't think this is the same as task state unfortently, so might need to make a new type for allocationState

```
\dT+ allocation_state 
                                               List of data types
 Schema |       Name       |  Internal name   | Size |  Elements   |  Owner   | Access privileges | Description 
--------+------------------+------------------+------+-------------+----------+-------------------+-------------
 public | allocation_state | allocation_state | 4    | PENDING    +| postgres |                   | 
        |                  |                  |      | ASSIGNED   +|          |                   | 
        |                  |                  |      | PULLING    +|          |                   | 
        |                  |                  |      | STARTING   +|          |                   | 
        |                  |                  |      | RUNNING    +|          |                   | 
        |                  |                  |      | WAITING    +|          |                   | 
        |                  |                  |      | TERMINATING+|          |                   | 
        |                  |                  |      | TERMINATED  |          |                   | 
(1 row)
```

File: proto/src/determined/api/v1/trial.proto [hunk 8]
```diff
@@ -520,6 +520,30 @@ message AllocationRendezvousInfoResponse {
   determined.trial.v1.RendezvousInfo rendezvous_info = 1;
 }
 
+// Gets some allocation.
```
Comment (NicholasBlaskey) [inline proto/src/determined/api/v1/trial.proto:523]: Maybe move to task.proto?

**Review by NicholasBlaskey — APPROVED** [review]
LGTM except for a couple comments

File: proto/src/determined/api/v1/trial.proto (hunk 8 above)
Comment (salonig23) [reply proto/src/determined/api/v1/trial.proto]: Done

**Pull request merged**